pkg/services/k8s/apiserver/admission/grpc_callout_admission_in_plugin.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/plugins"
+	admissionV1 "k8s.io/api/admission/v1"
+	authenticationV1 "k8s.io/api/authentication/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	runtimeJson "k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apiserver/pkg/admission"
+)
+
+const PluginNameGrpcCalloutAdmissionInPlugin = "GrpcCalloutAdmissionInPlugin"
+
+// Register registers a plugin
+func RegisterGrpcCalloutAdmissionInPlugin(plugins *admission.Plugins, pluginsClient plugins.Client) {
+	plugins.Register(PluginNameGrpcCalloutAdmissionInPlugin, func(config io.Reader) (admission.Interface, error) {
+		return NewGrpcCalloutAdmissionInPlugin(pluginsClient), nil
+	})
+}
+
+// example of admission plugin that will deny any resource with name "deny"
+type grpcCalloutAdmissionInPlugin struct {
+	pluginsClient plugins.Client
+}
+
+var _ admission.MutationInterface = grpcCalloutAdmissionInPlugin{}
+var _ admission.ValidationInterface = grpcCalloutAdmissionInPlugin{}
+
+// Admit makes an admission decision based on the request attributes.
+func (g grpcCalloutAdmissionInPlugin) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	return g.proxyAdminOrValidate(ctx, a, o, true)
+}
+
+// Validate makes an admission decision based on the request attributes.  It is NOT allowed to mutate.
+func (g grpcCalloutAdmissionInPlugin) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	return g.proxyAdminOrValidate(ctx, a, o, false)
+}
+
+func (g grpcCalloutAdmissionInPlugin) proxyAdminOrValidate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces, isMutating bool) error {
+	var finalErr error
+
+	wrappedSender := callResourceResponseSenderFunc(func(response *backend.CallResourceResponse) error {
+		// err = errors.New("some validation error")
+		admissionResponse := &admissionV1.AdmissionResponse{}
+		if err := json.Unmarshal(response.Body, &admissionResponse); err != nil {
+			finalErr = errors.New("Admission response from plugin is malformed")
```

Comment:
a few questions here:
* does the error need to be in a specific format? i've seen issues where returning normal errors aren't handled properly by k8s code that's expecting a specific type of error
* should this return after setting `finalErr`?
* what happens if you return an error in the wrapper function? i'm guessing that's not what we want, but i'm just curious how that behaves

Reply:
I can totally see that the error handling here is confusing and sub-par. 

* Re: specific format: so far, I haven't noticed that the error needs to be in a specific format. I will confirm what the HTTP status code I am seeing for generic string errors.
* should this return after setting finalErr? Yes (will add).
* Re: returning an error in the wrapper function. I have the same question. Its not something I want but I can check the behavior.